### ISSUE TITLE
npm:ractive-load

### ISSUE THREAD

**Issue by guybedford** [jspm/registry#115]
See https://github.com/jspm/npm/issues/39#issuecomment-70544375.

**Comment by guybedford** [comment]
@victorwpbastos following up from https://github.com/jspm/npm/issues/39#issuecomment-70589624:

I think you're jumping a few steps - the first thing is to just get `ractive-load` to work on its own.

Test out:

```
  jspm install npm:ractive-load -o {...}
```

With a simple page to load `System.import('ractive-load')`.

Then there will be an error, specifically I'm getting:

```
 Uncaught Error: Could not find Ractive! It must be loaded before the Ractive.load plugin
system.js:2161 Uncaught Error evaluating file:///Users/guybedford/Projects/jspm-cli/sandbox/jspm_packages/npm/ractive-load@0.4.0/ractive-load.js
Uncaught Error: Could not find Ractive! It must be loaded before the Ractive.load plugin
es6-module-loader.js:139 Potentially unhandled rejection [2] Error loading "npm:ractive-load@0.4.0" at file:///Users/guybedford/Projects/jspm-cli/sandbox/jspm_packages/npm/ractive-load@0.4.0.js
Error evaluating file:///Users/guybedford/Projects/jspm-cli/sandbox/jspm_packages/npm/ractive-load@0.4.0.js
Uncaught Error evaluating file:///Users/guybedford/Projects/jspm-cli/sandbox/jspm_packages/npm/ractive-load@0.4.0/ractive-load.js
Uncaught Error: Could not find Ractive! It must be loaded before the Ractive.load plugin (WARNING: non-Error used)
```

I don't have time to help with this, but when I do I will.

**Comment by victorwpbastos** [comment]
Yes, then I installed github:ractive, required it before ractive-load and everything works in runtime but in bundle not.

**Comment by guybedford** [comment]
@victorwpbastos ok I see. Yes, just requiring ractive before ractive-load is not enough, as you need to ensure the dependency is defined at a modular level. This is where you should use shim configuration to enforce the dependency https://github.com/jspm/registry/wiki/Configuring-Packages-for-jspm#shim-configuration.

**Comment by guybedford** [comment]
Also, perhaps use ractive-load from GitHub as well then?

**Comment by guybedford** [comment]
Basically, something like

```
  jspm install github:ractivejs/ractive-load -o "{ shim: { 'ractive-load': ['ractive'] }, dependencies: { ractive: 'github:ractivejs/ractive' } }"
```

**Comment by victorwpbastos** [comment]
Runtime: fine, bundle:

![image](https://cloud.githubusercontent.com/assets/1017515/5819700/ee5f2898-a0a6-11e4-84a0-753652b9654d.png)

Ract plugin:

```js
exports.translate = function(load) {
	load.source = 'module.exports = ' + JSON.stringify(load.source);
	return Ractive.load(load.address).then(function(c) {
		Ractive.components['ract-teste'] = c;
	});
};
```

**Comment by guybedford** [comment]
Yes this is a race condition that is only affecting you on the server as the dependency on `Ractive` is not properly defined.

**Comment by victorwpbastos** [comment]
So by now there is no way to do that?

**Comment by guybedford** [comment]
This will be fixed by applying the shim override like the above example in my previous comment.

**Comment by victorwpbastos** [comment]
After that I could install ractive-load running: `jspm install ractive-load` and jspm would take care of everything, is it?

**Comment by guybedford** [comment]
Yes exactly.

**Comment by victorwpbastos** [comment]
I wanna help. How could I do that?

**Comment by guybedford** [comment]
I'd suggest trying to follow the suggestion in https://github.com/jspm/registry/issues/115#issuecomment-70651199 and see where it breaks. Read about overrides in detail at https://github.com/jspm/registry/wiki/Configuring-Packages-for-jspm#shim-configuration.

**Comment by guilhermeaiolfi** [comment]
@victorwpbastos any luck with that? I'm having the exactly same problem. But I'm not using ractive-node but rcu. rcu is a dep used by ractive-load. @guybedford suggestion didn't worked.

**Comment by victorwpbastos** [comment]
I came back to webpack until this is resolved.
Em 08/04/2015 15:36, "guilhermeaiolfi" <notifications@github.com> escreveu:

**Comment by guilhermeaiolfi** [comment]
I wonder if that's something that @Rich-Harris could help and change in the library side.

**Comment by guilhermeaiolfi** [comment]
So I made some progress. Enforcing CJS as a format for rcu ( "format cjs" ) and instaling github:jspm/nodelibs-module and requiring it at the top of the file made it work until:

```
     Cannot call method 'add' of null
         at Object.cssConfigurator.extend (evalmachine.<anonymous>:3400:17)
         at configure (evalmachine.<anonymous>:6977:19)
         at Object.config.extend (evalmachine.<anonymous>:6932:12)
         at extendOne (evalmachine.<anonymous>:16531:18)
         at Array.reduce (native)
         at Function._extend__extend (evalmachine.<anonymous>:16491:20)
         at createComponent [as _onTimeout] (evalmachine.<anonymous>:387:27)
         at Timer.listOnTimeout [as ontimeout] (timers.js:110:15)
``` 
But that's OK, seems a library error, so commented the line in ractive@0.7.2:7400
```
global_css.add({ id: id, styles: styles });
```

then the build completes. Not a lot of changes to make. What you guys think?

**Comment by victorwpbastos** [comment]
In fact I'm using handlebars again. Ractive seems a little work in progress yet. I hope they keep the good work. Looks promising!

**Comment by guilhermeaiolfi** [comment]
The problem with ```global_css.add({ id: id, styles: styles });``` was my fault. The only problem left is enforcing the cjs format for rcu and requiring 'module' (nodelibs-module). Here is the code: https://github.com/guilhermeaiolfi/ractem

**Comment by guybedford** [comment]
@guilhermeaiolfi thanks for updating here. I've just added an experimental feature to jspm at https://github.com/jspm/jspm-cli/pull/670, which would allow shim to apply to any module format.

With this feature, such an install can be accomplished with:

```
jspm install npm:rcu -o "{shim:{'dist/rcu':{deps:['module']}},dependencies:{module:'jspm/nodelibs-module@0.1'},format:'cjs'}"
```

This is breaking though, so won't be released too soon unfortunately, but let me know if it looks like that will help here.

**Comment by guilhermeaiolfi** [comment]
So, it worked for me:

```javascript
System.config({
  meta: {
    'github:ractivejs/rcu@0.4.0/dist/rcu': {
      deps: ['module'],
      format: 'cjs'
    }
  }
});
```

It took me a while until I understood how 'meta' works because it matches files and not modules names (at least in this case).

DO NOT WORK:
```javascript
System.config({
  meta: {
    'rcu': { // 'github:ractivejs/rcu@0.4.0' won't work either
      deps: ['module'],
      format: 'cjs'
    }
  }
});
```
I hope it helps someone else. You can close this issue for now.

Thanks for the tips.
